src/inmanta/data/model.py

```diff
@@ -747,31 +749,48 @@ class LoginReturn(BaseModel):
     user: User
 
 
+def is_resource_id(v: Optional[str], info: ValidationInfo, allow_none: bool = False) -> str | None:
+    if v is None:
+        if allow_none:
+            return v
+        raise ValidationError(f"{info.field_name} is None")
+    return resources.Id.parse_id(v).resource_str()
+
+
 class DiscoveredResource(BaseModel):
     """
     :param discovered_resource_id: The name of the resource
     :param values: The actual resource
     :param managed_resource_uri: URI of the resource with the same ID that is already
         managed by the orchestrator e.g. "/api/v2/resource/<rid>". Or None if the resource is not managed.
+    :param discovery_resource_id: Resource id of the (managed) discovery resource that reported this
+        discovered resource.
     """
 
     discovered_resource_id: ResourceIdStr
     values: dict[str, object]
     managed_resource_uri: Optional[str] = None
 
-    @field_validator("discovered_resource_id")
-    @classmethod
-    def discovered_resource_id_is_resource_id(cls, v: str) -> str:
-        if resources.Id.is_resource_id(v):
-            return v
-        raise ValueError(f"id {v} is not of type ResourceIdStr")
+    # This field is excluded during serialization in favour of the discovery_resource_uri
+    discovery_resource_id: Optional[ResourceIdStr] = Field(default=None, exclude=True)
```

Comment:
@arnaudsjs tagging you because this is a deviation of what you reviewed.

Reply:
@Hugo-Inmanta out of my comments it's probably best to consider this one first, because it may invalidate some of the others.

Reply:
> Do I understand correctly that this is optional because we may currently have None values in the database?

Yes, and also wouldn't that be a breaking change ? Won't this break in all the places where this endpoint is called without a `discovery_resource_id` ?

Reply:
Technically yes, but the method is defined as an agent-server API, and we only really ever use autostarted agents. So client and server should always use the same code afaik.

Reply:
Went with your approach for the typing. And the following:
------
DB side:
`DiscoveredResource`: add an optional field `discovery_resource_id`
The value of this field drives the choice of model class for serialization:
- if None -> `m.DiscoveredResource`
- else -> `m.LinkedDiscoveredResource`
------
Model side:
Parent class `DiscoveredResource` remains moslty unchanged
Add child class `LinkedDiscoveredResource` that links to an existing discovery resource

Reply:
> The value of this field drives the choice of model class for serialization

I don't think this is required, and it adds a lot of verbosity. The `LinkedDiscoveredResource` I proposed is just to declare the requirement that the field be set. This is relevant for the inbound API. For the outbound API it doesn't make any sense to promise we'll set the field, maybe. We might as well not promise and use the generic `DiscoveredResource`.

We should probably create a follow-up ticket to consider how to phase out the unlinked resources.

Reply:
Follow up ticket -> https://github.com/inmanta/inmanta-core/issues/8004